File: node/src/spiral.js
class Spiral {
  traverse(grid) {
    grid = grid;
    let rowStart = 0;
    let rowEnd = grid.length - 1;
    let colStart = 0;
    let colEnd = grid[0].length - 1;
    let result = [];

    while (rowStart <= rowEnd && colStart <= colEnd) {
      // Traverse down along the left
      for (let row = rowStart; row <= rowEnd; row++) {
        result.push(grid[row][colStart]);
      }
      colStart++;

      // Traverse across the bottom to the right
      for (let col = colStart; col <= colEnd; col++) {
        result.push(grid[rowEnd][col]);
      }
      rowEnd--;

      // Traverse up along the right
      if (rowStart <= rowEnd) {
        for (let row = rowEnd; row >= rowStart; row--) {
          result.push(grid[row][colEnd]);
        }
        colEnd--;
      }

      // Traverse across to the left
      if (colStart <= colEnd) {
        for (let col = colEnd; col >= colStart; col--) {
          result.push(grid[rowStart][col]);
        }
        rowStart++;
      }
    }
    return result;
  }
}

module.exports = Spiral;

File: golang/spiral.go
package main

type Spiral struct {
}

func (s *Spiral) Traverse(grid [][]int) []int {
	rowStart := 0
	rowEnd := len(grid) - 1
	colStart := 0
	colEnd := len(grid[0]) - 1
	result := make([]int, 0)

	for rowStart <= rowEnd && colStart <= colEnd {
		// Traverse down along the left
		for row := rowStart; row <= rowEnd; row++ {
			result = append(result, grid[row][colStart])
		}
		colStart++

		// Traverse across the bottom to the right
		for col := colStart; col <= colEnd; col++ {
			result = append(result, grid[rowEnd][col])
		}
		rowEnd--

		// Traverse up along the right
		if rowStart <= rowEnd {
			for row := rowEnd; row >= rowStart; row-- {
				result = append(result, grid[row][colEnd])
			}
			colEnd--
		}

		// Traverse across to the left
		if colStart <= colEnd {
			for col := colEnd; col >= colStart; col-- {
				result = append(result, grid[rowStart][col])
			}
			rowStart++
		}
	}
	return result
}
